processing/src/main/java/org/apache/druid/frame/read/columnar/StringArrayFrameColumnReader.java

```diff
@@ -0,0 +1,186 @@
+        true
+    );
+
+    return new ColumnAccessorBasedColumn(frameCol);
+  }
+
+  @Override
+  public ColumnPlus readColumn(final Frame frame)
+  {
+    final Memory memory = frame.region(columnNumber);
+    validate(memory);
+
+    // String arrays always store multiple values per row.
+    assert isMultiValue(memory);
+    final long startOfStringLengthSection = getStartOfStringLengthSection(frame.numRows());
+    final long startOfStringDataSection = getStartOfStringDataSection(memory, frame.numRows());
+
+    final BaseColumn baseColumn = new StringArrayFrameColumn(
+        frame,
+        memory,
+        startOfStringLengthSection,
+        startOfStringDataSection
+    );
+
+    return new ColumnPlus(
+        baseColumn,
+        new ColumnCapabilitiesImpl().setType(ColumnType.STRING_ARRAY)
+                                    .setHasMultipleValues(false)
+                                    .setDictionaryEncoded(false),
+        frame.numRows()
+    );
+  }
+
+  private void validate(final Memory region)
+  {
+    // Check if column is big enough for a header
+    if (region.getCapacity() < StringFrameColumnWriter.DATA_OFFSET) {
+      throw DruidException.defensive("Column[%s] is not big enough for a header", columnNumber);
+    }
+
+    final byte typeCode = region.getByte(0);
+    if (typeCode != FrameColumnWriters.TYPE_STRING_ARRAY) {
+      throw DruidException.defensive(
+          "Column[%s] does not have the correct type code; expected[%s], got[%s]",
+          columnNumber,
+          FrameColumnWriters.TYPE_STRING_ARRAY,
+          typeCode
+      );
+    }
+  }
+
+  static boolean isMultiValue(final Memory memory)
+  {
+    return memory.getByte(StringFrameColumnWriter.MULTI_VALUE_POSITION) == StringFrameColumnWriter.MULTI_VALUE_BYTE;
+  }
+
+  private static long getStartOfCumulativeLengthSection()
+  {
+    return StringFrameColumnWriter.DATA_OFFSET;
+  }
+
+  private static long getStartOfStringLengthSection(final int numRows)
+  {
+    return StringFrameColumnWriter.DATA_OFFSET + (long) Integer.BYTES * numRows;
+  }
+
+  private long getStartOfStringDataSection(
+      final Memory memory,
+      final int numRows
+  )
+  {
+    if (numRows < 0) {
+      throw DruidException.defensive("Encountered -ve numRows [%s] while reading frame", numRows);
+    }
+    final int totalNumValues = FrameColumnReaderUtils.getAdjustedCumulativeRowLength(
+        memory,
+        getStartOfCumulativeLengthSection(),
+        numRows - 1
+    );
+
+    return getStartOfStringLengthSection(numRows) + (long) Integer.BYTES * totalNumValues;
+  }
+
+  private static class StringArrayFrameColumn implements BaseColumn
+  {
+    private final StringFrameColumnReader.StringFrameColumn delegate;
+
+    StringArrayFrameColumn(
+        Frame frame,
+        Memory memory,
+        long startOfStringLengthSection,
+        long startOfStringDataSection
+    )
+    {
+      this.delegate = new StringFrameColumnReader.StringFrameColumn(
```

Comment:
hmm, this seems like it should be split out into its own thing too... partially splitting out seems confusing to me. Part of the reason to split things up is so that the string frame column isn't trying to be 3 types at once and can remove all the branchy stuff like "am i multi value? and if so am i an array or regular mvd?". Also array types don't need to implement dimension selectors, which will never be asked for with a real array type, and so on.

I guess its fine if we want to do this as a follow-up, but this doesn't seem like a finished clean split, which is what I would like things to end up as so that everything is completely unambiguous what type we are dealing with and minimizing the chance that we mistakenly implicitly translate between these types.

Reply:
@clintropolis I think the better way is to make `StringArrayFrameColumn` extending `StringFrameColumn` rather than completely splitting out and renaming the confusing `multiValue` names to `hasMultipleVaues` as you suggested.
else it would be a complete copy and readability wise also we could remove those hard coded boolean values in the readers.

what do you think?

Reply:
Segregated completely